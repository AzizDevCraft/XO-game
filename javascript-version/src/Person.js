export default class Person {
    #name;
    #symbol;

    /**
     * @param {string} name 
     * @param {string} symbol 
     * @param {string} existingName 
     * @param {string} existingSymbol 
     */
    constructor (name, symbol, existingName="", existingSymbol="") {
        const nameValidation = Person.validName (name, existingName)
        this.#name = nameValidation.valid ? nameValidation.name : ""
        const symbolValidation = Person.validSymbol (symbol, existingSymbol)
        this.#symbol = symbolValidation.valid ? symbolValidation.symbol : ""
        this._score = 0
        this._game = 0
    }

    /**
     * @static
     * @param {string} name 
     * @param {string} existingName 
     * @returns {{valid:boolean, errorMessage?:string, name?:string}}
     */
    static validName (name, existingName) {
        if (name.trim ().toLowerCase () === existingName.trim ().toLowerCase ()) { 
            return {valid : false, errorMessage : "nom existe deja !"} 
        } 
        else if (!isNaN(Number.parseInt(name))) {
            return {valid : false, errorMessage : "nom invalide ! on peut pas mettre que des chiffres ou commencer avec !"}
        } 
        else if (name.length < 4){
            return {valid : false, errorMessage : "nom invalide (une longueur minimum de 4 lettre) !"}
        }
        else return {valid : true, name}
    }

    /**
     * @static
     * @param {string} symbol 
     * @param {string} existingSymbol 
     * @returns {{valid:boolean, errorMessage?:string, symbol?:string}}
     */
    static validSymbol (symbol, existingSymbol) {
        if (symbol.trim ().toLowerCase () === existingSymbol.trim ().toLowerCase ()) {
            return {valid : false, errorMessage : "symbol existe deja !"}            
        }
        else if (!isNaN(Number (symbol))) {
            return {valid : false, errorMessage : "Syntaxe invalid (pas de chiffre) !"}            
        }
        else if (symbol.length !== 1){
            return {valid : false, errorMessage : "il faut que la longeur égale à 1 !"}            
        }
        else return {valid : true, symbol}
    }

    get _name () {
        return this.#name[0].toUpperCase () + this.#name.slice (1).toLowerCase ()
    }

    get _symbol () {
        return this.#symbol
    }

    /**
     * @param {string} newSymbol 
     * @param {string} existingSymbol 
     */
    _updateSymbol (newSymbol, existingSymbol) {
        if (Person.validSymbol (newSymbol, existingSymbol).valid) {
            this.#symbol = newSymbol
        }
        else throw new SyntaxError ("Symbol invalide")
    }
}