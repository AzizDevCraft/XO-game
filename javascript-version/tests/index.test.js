import GameController from "../src/index.js"
import FakeIO from "./FakeIO"
import Person from "../src/Person.js"

describe ("tester la validation des joueurs", () => {

    describe ("tester askUntilValidPerson", () => {
        test ("accepter un nom valide", () => {
            const IO = new FakeIO (["aziz"])
            const game = new GameController (IO)
            expect (game.askUntilValidPerson ("Donner votre nom : ", Person.validName, "")).toMatch ("aziz")
        })
        
        test ("refuser jusqu'à avoir un nom valid", () => {
            const IO = new FakeIO (["AZ", "1234", "12Aziz", "Ouss", "aziz", "Foulen"])
            const game = new GameController (IO)
            expect (game.askUntilValidPerson ("Donner votre nom : ", Person.validName, "Ouss")).toMatch ("aziz")
        })

        test ("accepter un symbole valide", () => {
            const IO = new FakeIO (["X"])
            const game = new GameController (IO)
            expect (game.askUntilValidPerson ("choisissez votre symbol : ", Person.validSymbol, "")).toMatch ("X")
        })
        
        test ("refuser jusqu'à avoir un symbole valid", () => {
            const IO = new FakeIO (["1", "AA", "X", "N", "O"])
            const game = new GameController (IO)
            expect (game.askUntilValidPerson ("choisissez votre symbol : ", Person.validSymbol, "X")).toMatch ("N")
        })
    })
    
    describe ("tester le debut du jeux", () => {
        test ("Happy Path", () => {
            const IO = new FakeIO ([1, "aziz", "X", "OUSS", "O"])
            const game = new GameController (IO)
            game.startGame ()

            expect (IO.output[0]).toContain ("Bienvenue à X - O game")

            const player = game.players[0]
            expect (player._name).toMatch("Aziz")
            expect (player._symbol).toMatch("X")
            expect (IO.inputs).toHaveLength (0)
            expect (IO.output).toHaveLength (8)

            expect (game.game.players).toEqual (game.players)
        })

        test ("refuser identique players", () => {
            const IO = new FakeIO ([1, "aziz", "X", "Aziz", "AZIZ", "OUss", "O"]) 
            const game = new GameController (IO)
            game.startGame ()
            const player2 = game.players [1]
            expect (player2._name).not.toMatch("Aziz")
            expect (IO.output).toHaveLength (10)
        })

        test ("refuser un nom non valide", () => {
            const IO = new FakeIO ([1, "AB", "12345", "12AZIZ", "Zizz123", "X", "Ouss", "O"]) 
            const game = new GameController (IO)
            game.startGame ()
            const player1 = game.players[0]
            expect (player1._name).toMatch("Zizz123")
            expect (player1._symbol).toMatch("X")
        })
        
        test ("refuser un symbol non valide", () => {
            const IO = new FakeIO ([1, "Aziz", "0", "ABC", "X", "Ouss", "X", "X", "O"]) 
            const game = new GameController (IO)
            game.startGame ()
            expect (game.players[0]._symbol).toMatch("X")
            expect (game.players[1]._symbol).toMatch("O")
        })

        test ("verifier le deuxième choix (choix = 2)", () => {
            const IO = new FakeIO ([2]) 
            const game = new GameController (IO)
            game.startGame ()
            expect (IO.output).toContain ("Game Over")
            expect (IO.output.at (-1)).toMatch ("console fermer")
        })
        
        test ("saisi de choix sous forme de String", () => {
            const IO = new FakeIO (["2"]) 
            const game = new GameController (IO)
            game.startGame ()
            expect (IO.output).toContain ("Game Over")
            expect (IO.output.at (-1)).toMatch ("console fermer")
        })
        
        test ("refuser un choix invalide", () => {
            const IO = new FakeIO (["3", 4, 2]) 
            const game = new GameController (IO)
            game.startGame ()
            expect (IO.output.at (-1)).toMatch ("console fermer")
        })
    })

})

describe ("tester la partie", () => {

    describe ("tester la methode principale playGame", () => {
        let IO;
        let game;

        beforeEach (() => {
            IO = new FakeIO ([1, "aziz", "X", "OUSS", "O"])
            game = new GameController (IO)
            game.startGame ()   
        })
        
        test ("win moves for player 1 et la boucle s'arrête au bon moment", () => {
            IO.inputs.push ("1", "2", "3", "4", "5", "6", "7", "8")
            game.playGame ()
            expect (IO.output).toMatchSnapshot ()
        })
        
        test ("tester le cas de draw", () => {
            IO.inputs.push ("1", "2", "3", "5", "4", "6", "8", "7", "9")
            game.playGame ()
            expect (IO.output).toMatchSnapshot ()
        })
        
        test ("Coup invalid", () => {
            IO.inputs.push ("1", "2", "2", "5", "4", "6", "8", "7", "9")
            expect (() => game.playGame ()).toThrow ()
        })
    })

    describe ("tester restartGame", () => {
        test ("tester l'affichage du menu de fin", () => {
            const IO = new FakeIO ([2])
            const game = new GameController (IO)
            game.restartGame ()
            expect (IO.output[0]).toContain ("Game Over !")
        })
        
        test ("valider le premier choix avec un cas de win ", () => {
            const IO = new FakeIO ([1, "aziz", "X", "OUSS", "O", 1, "1", "2", "3", "4", "5", "6", "7"])
            const game = new GameController (IO)
            game.startGame ()
            game.restartGame ()
            expect (IO.output).toMatchSnapshot ()
        })
        
        test ("valider le deuxième choix", () => {
            const IO = new FakeIO ([2])
            const game = new GameController (IO)
            game.restartGame ()
            expect (IO.output.at(-1)).toMatch ("console fermer")
        })
    })


})