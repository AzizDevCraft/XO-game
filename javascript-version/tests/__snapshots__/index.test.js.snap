// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`tester la partie tester la methode principale playGame tester le cas de draw 1`] = `
[
  "
Bienvenue à X - O game
  1. Start the game 
  2. Quit the game",
  "choisissez une option : ",
  "player 1 identifiez-vous :",
  "Donner votre nom : ",
  "choisissez votre symbol : ",
  "player 2 identifiez-vous :",
  "Donner votre nom : ",
  "choisissez votre symbol : ",
  " 1 | 2 | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | 2 | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | O | X 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | X 
___|___|___
 4 | O | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | O | X 
___|___|___
 X | O | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | X 
___|___|___
 X | O | O 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | O | X 
___|___|___
 X | O | O 
___|___|___
 7 | X | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | X 
___|___|___
 X | O | O 
___|___|___
 O | X | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | O | X 
___|___|___
 X | O | O 
___|___|___
 O | X | X 
   |   |   ",
  "égalité !",
]
`;

exports[`tester la partie tester la methode principale playGame win moves for player 1 et la boucle s'arrête au bon moment 1`] = `
[
  "
Bienvenue à X - O game
  1. Start the game 
  2. Quit the game",
  "choisissez une option : ",
  "player 1 identifiez-vous :",
  "Donner votre nom : ",
  "choisissez votre symbol : ",
  "player 2 identifiez-vous :",
  "Donner votre nom : ",
  "choisissez votre symbol : ",
  " 1 | 2 | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | 2 | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | O | X 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | X 
___|___|___
 O | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | O | X 
___|___|___
 O | X | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | X 
___|___|___
 O | X | O 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  "Aziz a gagné cette partie !",
  " X | O | X 
___|___|___
 O | X | O 
___|___|___
 X | 8 | 9 
   |   |   ",
]
`;

exports[`tester la partie tester restartGame valider le premier choix avec un cas de win  1`] = `
[
  "
Bienvenue à X - O game
  1. Start the game 
  2. Quit the game",
  "choisissez une option : ",
  "player 1 identifiez-vous :",
  "Donner votre nom : ",
  "choisissez votre symbol : ",
  "player 2 identifiez-vous :",
  "Donner votre nom : ",
  "choisissez votre symbol : ",
  "
Game Over !
  1. Restart Game
  2. Quit Game",
  "choisissez une option : ",
  " 1 | 2 | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | 2 | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | 3 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | O | X 
___|___|___
 4 | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | X 
___|___|___
 O | 5 | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  " X | O | X 
___|___|___
 O | X | 6 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Ouss, choisit une position : ",
  " X | O | X 
___|___|___
 O | X | O 
___|___|___
 7 | 8 | 9 
   |   |   ",
  "c'est le tour de Aziz, choisit une position : ",
  "Aziz a gagné cette partie !",
  " X | O | X 
___|___|___
 O | X | O 
___|___|___
 X | 8 | 9 
   |   |   ",
]
`;
